### PR TITLE
Add OTel attributes to Event Hubs library

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Contributes to https://github.com/Azure/azure-sdk-for-net/issues/31650

### CODE REVIEW

File: sdk/core/Azure.Core/src/Shared/MessagingClientDiagnostics.cs [hunk 1]
```diff
@@ -148,24 +156,56 @@ public static bool TryExtractTraceContext(IDictionary<string, object> properties
...
+                        if (activity.TraceStateString != null)
+                        {
+                            tracestate = activity.TraceStateString;
+                            properties[TraceState] = tracestate;
+                        }
+                    }
                 }
             }
+            else
+                TryExtractTraceContext(properties, out traceparent, out tracestate);
```
Comment (jsquire) [inline sdk/core/Azure.Core/src/Shared/MessagingClientDiagnostics.cs:193]:
```suggestion
            else
            {
                TryExtractTraceContext(properties, out traceparent, out tracestate);
            }
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs [hunk 2]
```diff
@@ -340,6 +341,11 @@ public new bool IsRunning
         ///
         private CheckpointStore CheckpointStore { get; }
 
+        /// <summary>
+        ///   The client diagnostics for this processor.
+        /// </summary>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs:346]:
```suggestion
        /// </summary>
        ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    ///   marked as non-parallelizable.
+    /// </remarks>
+    ///
+    [NonParallelizable]
+    [TestFixture]
+    public class DiagnosticsActivitySourceTests
+    {
+        /// <summary>
+        /// Resets the activity source feature switch after each test.
+        /// </summary>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs:45]:
```suggestion
        /// <summary>
        ///   Resets the activity source feature switch after each test.
        /// </summary>
        ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        ///
+        /// <param name="target">The client whose method to invoke.</param>
+        /// <param name="partitionId">The identifier of the partition the checkpoint is for.</param>
+        /// <param name="offset">The offset to associate with the checkpoint, indicating that a processor should begin reading form the next event in the stream.</param>
+        /// <param name="sequenceNumber">An optional sequence number to associate with the checkpoint, intended as informational metadata.  The <paramref name="offset" /> will be used for positioning when events are read.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> instance to signal a request to cancel the operation.</param>
+        ///
+        /// <returns>The translated options.</returns>
+        ///
+        private static Task InvokeUpdateCheckpointAsync(EventProcessorClient target,
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs:107]: It's a protected method; why not extend with a test shim?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs [hunk 5]
```diff
@@ -225,6 +226,11 @@ internal EventProcessorStatus Status
         ///
         private int EventBatchMaximumCount { get; }
 
+        /// <summary>
+        ///   The client diagnostics for this processor.
+        /// </summary>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor.cs:231]:
```suggestion
        /// <summary>
        ///   The client diagnostics for this processor.
        /// </summary>
        ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventDataBatch.cs [hunk 6]
```diff
@@ -173,17 +181,17 @@ public bool TryAdd(EventData eventData)
                 AssertNotLocked();
 
                 var messageScopeCreated = false;
-                var identifier = default(string);
+                var traceparent = default(string);
 
                 try
                 {
-                    (messageScopeCreated, identifier) = EventDataInstrumentation.InstrumentEvent(eventData, FullyQualifiedNamespace, EventHubName);
+                    ClientDiagnostics.InstrumentMessage(eventData.Properties, DiagnosticProperty.EventActivityName, out messageScopeCreated, out traceparent, out var tracestate);
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventDataBatch.cs:188]: Why all the out params?   I find that much less clear than deconstructing a tuple on return.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs [hunk 7]
```diff
@@ -150,6 +151,11 @@ public bool IsClosed
         ///
         private ConcurrentDictionary<string, PartitionPublishingState> PartitionState { get; }
 
+        /// <summary>
+        ///   The client diagnostics for this producer.
+        /// </summary>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs:156]:
```suggestion
        /// <summary>
        ///   The client diagnostics for this producer.
        /// </summary>
        ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Amqp/AmqpProducerTests.cs [hunk 8]
```diff
@@ -1591,7 +1597,7 @@ public async Task SendBatchCreatesTheAmqpMessageFromTheBatch(string partitionKey
 
             using TransportEventBatch transportBatch = await producer.Object.CreateBatchAsync(options, default);
 
-            using var batch = new EventDataBatch(transportBatch, "ns", "eh", options);
+            using var batch = new EventDataBatch(transportBatch, "ns", "eh", options, new MessagingClientDiagnostics("mock", "mock", "mock", "mock", "mock"));
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Amqp/AmqpProducerTests.cs:1600]: Why not use the `MockClientDiagnotics` instance here and on L1653?

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsActivitySourceTests.cs [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+    ///   marked as non-parallelizable.
+    /// </remarks>
+    ///
+    [NonParallelizable]
+    [TestFixture]
+    public class DiagnosticsActivitySourceTests
+    {
+        /// <summary>
+        /// Resets the activity source feature switch after each test.
+        /// </summary>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsActivitySourceTests.cs:45]:
```suggestion
        /// <summary>
        ///   Resets the activity source feature switch after each test.
        /// </summary>
        ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsActivitySourceTests.cs [hunk 10]
```diff
@@ -0,0 +1,585 @@
...
+
+            if (eventCount > 1)
+                CollectionAssert.Contains(tags, new KeyValuePair<string, int>(MessagingClientDiagnostics.BatchCount, eventCount));
+            else
+                CollectionAssert.DoesNotContain(tags, new KeyValuePair<string, int>(MessagingClientDiagnostics.BatchCount, eventCount));
+        }
+
+        /// <summary>
+        /// Sets and returns the app config switch to enable Activity Source. The switch must be disposed at the end of the test.
+        /// </summary>
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/tests/Diagnostics/DiagnosticsActivitySourceTests.cs:555]:
```suggestion
        /// <summary>
        ///   Sets and returns the app config switch to enable Activity Source. The switch must be disposed at the end of the test.
        /// </summary>
        ///
```

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs:107]: Hmm this was copied verbatim from [DiagnosticTests](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsTests.cs#L89).

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs:107]: Then I'm a worthless hack that forgot to update after we changed to `protected` there.   Feel free to leave as is and I'll bask in my shame until I fix both.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventDataBatch.cs (hunk 6 above)
Comment (JoshLove-msft) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventDataBatch.cs:188]: I thought it was more appropriate given that not all callers care about the return value. The out params represent extra info that you can get back as opposed to being the primary reason for calling the method. That said, I can change it back if you prefer.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs:107]: That was my initial thought (minus the worthless hack part), but then I saw this [line](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsTests.cs#L78).

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/Diagnostics/DiagnosticsActivitySourceTests.cs:108]: ... then I have no idea what past me was thinking, but I assume he is smarter than present me.

**Review by jsquire — APPROVED** [review]
If you'd be so kind, let's make sure the live tests pass consistently before we merge.   The SB suite has been falling since the equivalent change over there.

**Pull request merged**